File: FNM_RE/modules/toc_structure.py
"""阶段 1 模块：目录结构与章节骨架。"""

from __future__ import annotations

from dataclasses import asdict
from typing import Any, Mapping

from FNM_RE.modules.contracts import GateReport, ModuleResult
from FNM_RE.modules.types import TocChapter, TocNode, TocPageRole, TocSectionHead, TocStructure
from FNM_RE.shared.title import chapter_title_match_key, normalize_title
from FNM_RE.stages.chapter_skeleton import build_chapter_skeleton
from FNM_RE.stages.page_partition import build_page_partitions, summarize_page_partitions
from FNM_RE.stages.section_heads import build_section_heads
from FNM_RE.shared.notes import _safe_int

_BACK_MATTER_REASON_HINTS = {
    "appendix",
    "bibliography",
    "index",
    "illustrations",
    "rear_toc_tail",
    "rear_author_blurb",
    "rear_sparse_other",
}

def _normalize_toc_node_role(value: Any) -> str:
    role = str(value or "").strip().lower().replace("-", "_")
    if role in {"container", "endnotes", "chapter", "section", "post_body", "back_matter", "front_matter"}:
        return role
    if role == "part":
        return "container"
    if role in {"notes", "endnotes"}:
        return "endnotes"
    if role == "book_title":
        return "front_matter"
    return ""

def _map_toc_role(
    title: str,
    *,
    explicit_role: str,
    title_key: str,
    container_keys: set[str],
    endnotes_keys: set[str],
    post_body_keys: set[str],
    back_matter_keys: set[str],
    chapter_keys: set[str],
) -> str:
    normalized_explicit_role = _normalize_toc_node_role(explicit_role)
    if normalized_explicit_role:
        return normalized_explicit_role
    if title_key in container_keys:
        return "container"
    if title_key in endnotes_keys:
        return "endnotes"
    if title_key in post_body_keys:
        return "post_body"
    if title_key in back_matter_keys:
        return "back_matter"
    if title_key in chapter_keys:
        return "chapter"
    normalized = normalize_title(title).lower()
    if "endnote" in normalized or normalized == "notes" or normalized.startswith("notes "):
        return "endnotes"
    if any(token in normalized for token in ("bibliograph", "index", "appendix", "references")):
        return "back_matter"
    return "chapter"

def _build_toc_tree(
    toc_items: list[dict] | None,
    *,
    chapter_rows: list[TocChapter],
    meta: Mapping[str, Any],
) -> list[TocNode]:
    normalized_toc_rows = list(meta.get("normalized_toc_rows") or [])
    title_keys = {chapter_title_match_key(row.title) for row in chapter_rows}
    container_keys = {chapter_title_match_key(title) for title in list(meta.get("container_titles") or [])}
    endnotes_keys = {chapter_title_match_key(title) for title in list(meta.get("endnotes_titles") or [])}
    post_body_keys = {chapter_title_match_key(title) for title in list(meta.get("post_body_titles") or [])}
    back_matter_keys = {chapter_title_match_key(title) for title in list(meta.get("back_matter_titles") or [])}
    nodes: list[TocNode] = []
    source_rows = normalized_toc_rows if normalized_toc_rows else list(toc_items or [])
    for index, item in enumerate(source_rows, start=1):
        title = normalize_title(str(item.get("title") or ""))
        if not title:
            continue
        title_key = chapter_title_match_key(title)
        explicit_role = str(
            item.get("semantic_role")
            or item.get("role_hint")
            or item.get("explicit_role_hint")
            or ""
        )
        role = _map_toc_role(
            title,
            explicit_role=explicit_role,
            title_key=title_key,
            container_keys=container_keys,
            endnotes_keys=endnotes_keys,
            post_body_keys=post_body_keys,
            back_matter_keys=back_matter_keys,
            chapter_keys=title_keys,
        )
        target_pdf_page = _safe_int(item.get("target_pdf_page") or item.get("page_no") or 0)
        nodes.append(
            TocNode(
                node_id=str(item.get("item_id") or f"toc-node-{index:04d}"),
                title=title,
                role=role,  # type: ignore[arg-type]
                level=max(1, _safe_int(item.get("level") or 1)),
                target_pdf_page=target_pdf_page,
                parent_id=str(item.get("parent_id") or item.get("parent_title") or ""),
            )
        )
    return nodes

def _build_chapters(chapters: list[Any], *, meta: Mapping[str, Any]) -> list[TocChapter]:
    post_body_keys = {chapter_title_match_key(title) for title in list(meta.get("post_body_titles") or [])}
    rows: list[TocChapter] = []
    for row in chapters:
        title = normalize_title(str(getattr(row, "title", "") or ""))
        role = "post_body" if chapter_title_match_key(title) in post_body_keys else "chapter"
        rows.append(
            TocChapter(
                chapter_id=str(getattr(row, "chapter_id", "") or ""),
                title=title,
                start_page=_safe_int(getattr(row, "start_page", 0)),
                end_page=_safe_int(getattr(row, "end_page", 0)),
                pages=[_safe_int(page_no) for page_no in list(getattr(row, "pages", []) or []) if _safe_int(page_no) > 0],
                role=role,  # type: ignore[arg-type]
                source=str(getattr(row, "source", "fallback") or "fallback"),  # type: ignore[arg-type]
                boundary_state=str(getattr(row, "boundary_state", "ready") or "ready"),  # type: ignore[arg-type]
            )
        )
    rows.sort(key=lambda item: (item.start_page, item.chapter_id))
    return rows

def _build_page_roles(
    partitions: list[Any],
    *,
    chapters: list[TocChapter],
    back_matter_start_hint: int = 0,
) -> list[TocPageRole]:
    chapter_by_page: dict[int, TocChapter] = {}
    for chapter in chapters:
        for page_no in chapter.pages:
            chapter_by_page[int(page_no)] = chapter
        if chapter.start_page > 0 and chapter.end_page >= chapter.start_page:
            for page_no in range(chapter.start_page, chapter.end_page + 1):
                chapter_by_page.setdefault(page_no, chapter)

    first_chapter_start = min((int(ch.start_page) for ch in chapters if int(ch.start_page) > 0), default=0)
    total_pages = max(1, len(partitions))
    rear_page_role_min_page = max(24, int(total_pages * 0.45))
    
    if int(back_matter_start_hint or 0) > 0:
        back_matter_start = int(back_matter_start_hint)
    else:
        inferred_back_matter_start = min(
            (
                int(row.page_no)
                for row in partitions
                if str(row.reason or "") in _BACK_MATTER_REASON_HINTS 
                and int(row.page_no) > first_chapter_start
                and int(row.page_no) >= rear_page_role_min_page
            ),
            default=0,
        )
        back_matter_start = int(inferred_back_matter_start)

    rows: list[TocPageRole] = []
    for row in partitions:
        page_no = int(row.page_no)
        source_role = str(row.page_role or "").strip().lower()
        chapter = chapter_by_page.get(page_no)
        # 工单 #5：上游 note_scan（page_partition._rule_note_scan）已识别为
        # NOTES 容器页时，note 角色优先于 chapter，避免章末 NOTES 容器被
        # chapter mapping 覆盖。chapter_id 仍保留供下游 region 绑定。
        if source_role == "note":
            role = "note"
            chapter_id = chapter.chapter_id if chapter is not None else ""
        elif chapter is not None:
            role = chapter.role
            chapter_id = chapter.chapter_id
        elif page_no > 0 and back_matter_start > 0 and page_no >= back_matter_start:
            role = "back_matter"
            chapter_id = ""
        elif back_matter_start == 0 and str(row.reason or "") in _BACK_MATTER_REASON_HINTS and page_no >= rear_page_role_min_page:
            role = "back_matter"
            chapter_id = ""
        elif source_role in {"other"}:
            role = "front_matter"
            chapter_id = ""
        elif first_chapter_start > 0 and page_no < first_chapter_start:
            role = "front_matter"
            chapter_id = ""
        elif source_role == "front_matter":
            role = "front_matter"
            chapter_id = ""
        else:
            role = "front_matter"
            chapter_id = ""
        rows.append(
            TocPageRole(
                page_no=page_no,
                role=role,  # type: ignore[arg-type]
                source_role=str(row.page_role or ""),
                reason=str(row.reason or ""),
                chapter_id=chapter_id,
            )
        )
    rows.sort(key=lambda item: item.page_no)
    return rows

def _role_semantics_valid(pages: list[TocPageRole]) -> bool:
    first_back_matter_page = min((row.page_no for row in pages if row.role == "back_matter"), default=0)
    if first_back_matter_page <= 0:
        return True
    return not any(row.role == "chapter" and row.page_no > first_back_matter_page for row in pages)

def _chapter_order_monotonic(chapters: list[TocChapter], toc_tree: list[TocNode]) -> bool:
    chapter_targets = [
        int(node.target_pdf_page)
        for node in toc_tree
        if node.role in {"chapter", "post_body"} and int(node.target_pdf_page) > 0
    ]
    if len(chapter_targets) >= 2:
        return all(chapter_targets[index - 1] <= chapter_targets[index] for index in range(1, len(chapter_targets)))
    return all(chapters[index - 1].start_page <= chapters[index].start_page for index in range(1, len(chapters)))

def build_toc_structure(
    pages: list[dict],
    toc_items: list[dict] | None,
    *,
    manual_page_overrides: Mapping[str, Mapping[str, Any]] | None = None,
    pdf_path: str = "",
    visual_toc_bundle: Mapping[str, Any] | None = None,
) -> ModuleResult[TocStructure]:
    page_partitions = build_page_partitions(pages, page_overrides=manual_page_overrides)
    heading_candidates, phase1_chapters, chapter_meta = build_chapter_skeleton(
        page_partitions,
        toc_items=toc_items,
        toc_offset=0,
        pdf_path=str(pdf_path or ""),
        pages=pages,
        visual_toc_bundle=visual_toc_bundle,
    )
    section_heads, heading_review_summary = build_section_heads(
        phase1_chapters,
        heading_candidates,
        page_partitions,
        fallback_sections=list(chapter_meta.get("fallback_sections") or []),
    )

    toc_chapters = _build_chapters(phase1_chapters, meta=chapter_meta)
    toc_tree = _build_toc_tree(toc_items, chapter_rows=toc_chapters, meta=chapter_meta)
    toc_pages = _build_page_roles(
        page_partitions,
        chapters=toc_chapters,
        back_matter_start_hint=int(chapter_meta.get("back_matter_start_page") or 0),
    )
    toc_section_heads = [
        TocSectionHead(
            section_head_id=str(row.section_head_id or ""),
            chapter_id=str(row.chapter_id or ""),
            title=str(row.title or ""),
            page_no=int(row.page_no),
            level=int(row.level),
            source=str(row.source or ""),
        )
        for row in section_heads
    ]

    hard = {
        "toc.pages_classified": bool(toc_pages) and all(
            row.role in {"front_matter", "chapter", "post_body", "back_matter", "endnotes", "container"} for row in toc_pages
        ),
        "toc.has_exportable_chapters": any(row.role == "chapter" for row in toc_chapters),
        "toc.chapter_titles_aligned": bool(chapter_meta.get("chapter_title_alignment_ok", True)),
        "toc.chapter_order_monotonic": _chapter_order_monotonic(toc_chapters, toc_tree),
        "toc.role_semantics_valid": bool(chapter_meta.get("toc_semantic_contract_ok", True))
        and _role_semantics_valid(toc_pages),
    }
    soft = {
        "toc.section_alignment_warn": bool(chapter_meta.get("chapter_section_alignment_ok", True)),
        "toc.visual_toc_conflict_warn": int(chapter_meta.get("visual_toc_conflict_count") or 0) == 0,
    }
    reasons: list[str] = []
    if not hard["toc.pages_classified"]:
        reasons.append("toc_pages_unclassified")
    if not hard["toc.has_exportable_chapters"]:
        reasons.append("toc_no_exportable_chapter")
    if not hard["toc.chapter_titles_aligned"]:
        reasons.append("toc_chapter_title_mismatch")
    if not hard["toc.chapter_order_monotonic"]:
        reasons.append("toc_chapter_order_non_monotonic")
    if not hard["toc.role_semantics_valid"]:
        reasons.append("toc_role_semantics_invalid")

    overrides_used = [
        {"kind": "page_override", "page_no": int(row.page_no), "reason": "manual_override"}
        for row in page_partitions
        if str(row.reason or "") == "manual_override"
    ]
    evidence = {
        "page_partition_summary": summarize_page_partitions(page_partitions),
        "chapter_count": len(toc_chapters),
        "exportable_chapter_count": sum(1 for row in toc_chapters if row.role == "chapter"),
        "heading_graph_summary": dict(chapter_meta.get("heading_graph_summary") or {}),
        "toc_role_summary": dict(chapter_meta.get("toc_role_summary") or {}),
    }
    diagnostics = {
        "heading_review_summary": heading_review_summary,
        "heading_graph_summary": dict(chapter_meta.get("heading_graph_summary") or {}),
        "chapter_source_summary": dict(chapter_meta.get("chapter_source_summary") or {}),
        "container_titles": list(chapter_meta.get("container_titles") or []),
        "endnotes_titles": list(chapter_meta.get("endnotes_titles") or []),
        "post_body_titles": list(chapter_meta.get("post_body_titles") or []),
        "back_matter_titles": list(chapter_meta.get("back_matter_titles") or []),
        "endnote_explorer_hints": dict(chapter_meta.get("endnote_explorer_hints") or {}),
        "heading_candidates": list(heading_candidates or []),
        "chapter_meta": dict(chapter_meta or {}),
    }

    gate_report = GateReport(
        module="toc",
        hard=hard,
        soft=soft,
        reasons=reasons,
        evidence=evidence,
        overrides_used=list(overrides_used),
    )
    data = TocStructure(
        pages=toc_pages,
        toc_tree=toc_tree,
        chapters=toc_chapters,
        section_heads=toc_section_heads,
    )
    return ModuleResult(
        data=data,
        gate_report=gate_report,
        evidence=evidence,
        overrides_used=list(overrides_used),
        diagnostics={
            **diagnostics,
            "chapter_meta": dict(chapter_meta),
            "first_three_chapters": [asdict(row) for row in toc_chapters[:3]],
        },
    )

File: fnm_re_rs/fnm-phase1/src/toc_structure.rs
//! ←→ FNM_RE/modules/toc_structure.py
//! Phase1 顶层编排：组装所有子模块输出为 Phase1Structure。

use crate::chapter_skeleton::builder::build_chapter_skeleton;
use crate::heading_graph::build_heading_graph;
use crate::input::{ManualPageOverride, RawPage, TocItem};
use crate::page_partition::build_page_partitions;
use crate::page_roles::build_page_roles;
use crate::section_heads::build_section_heads;
use crate::toc_tree::build_toc_tree;
use fnm_core::db::Repository;
use fnm_core::records::{HeadingCandidate, Phase1Structure};
use fnm_core::types::ChapterSource;
use std::collections::HashMap;

/// TOC 门控报告，与 Python `GateReport` 对应。
/// ←→ FNM_RE/modules/contracts.py GateReport
#[derive(Debug, Clone, serde::Serialize)]
pub struct GateReport {
    pub hard: HashMap<String, bool>,
    pub soft: HashMap<String, bool>,
}

/// 已应用的覆写记录。
#[derive(Debug, Clone, serde::Serialize)]
pub struct OverrideRecord {
    pub kind: String,
    pub description: String,
}

#[derive(Debug, Clone)]
pub struct Phase1Config {
    pub manual_page_overrides: Option<HashMap<String, ManualPageOverride>>,
    /// VisualTOC bundle 中的 endnotes_start_page，用于 page_partition 的 front_matter 延续页修复。
    /// ←→ Python `build_toc_structure` 的 `endnotes_start_page`
    pub endnotes_start_page: Option<i64>,
    pub pdf_path: Option<String>,
    pub doc_id: Option<String>,
    /// LLM book-type 校验开关。当前 Rust 端 LLM 客户端尚未接入 phase1
    /// 主入口，默认 `true`（跳过）；传 `false` 触发 `anyhow::bail!`。
    pub skip_llm_verify: bool,
}

impl Default for Phase1Config {
    fn default() -> Self {
        Self {
            manual_page_overrides: None,
            endnotes_start_page: None,
            pdf_path: None,
            doc_id: None,
            skip_llm_verify: true,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Phase1Output {
    pub structure: Phase1Structure,
    pub diagnostics: serde_json::Value,
    /// 门控报告：硬 gate → 阻断管线，软 gate → 告警。
    pub gate_report: GateReport,
    /// 已应用的手动覆写记录。
    pub overrides_used: Vec<OverrideRecord>,
}

/// 检测 TOC 条目是否包含 OCR 乱码（控制字符比例 ≥5% 的条目占 ≥30%）。
/// ←→ FNM_RE/modules/toc_structure.py `_toc_items_look_garbled`
fn toc_items_look_garbled(toc_items: &[TocItem]) -> bool {
    if toc_items.is_empty() {
        return false;
    }
    let garbled_count = toc_items
        .iter()
        .filter(|item| {
            let title = &item.title;
            if title.is_empty() {
                return false;
            }
            let ctrl_count = title.chars().filter(|c| c.is_control()).count();
            ctrl_count > 0 && ctrl_count as f64 / title.len() as f64 > 0.05
        })
        .count();
    garbled_count > 0 && garbled_count as f64 >= toc_items.len() as f64 * 0.3
}

/// Phase1 主入口：从原始页面构建章节骨架。
pub fn build_phase1_structure(
    pages: &[RawPage],
    toc_items: Option<&[TocItem]>,
    config: &Phase1Config,
) -> anyhow::Result<Phase1Output> {
    let total_pages = pages.len() as i64;

    // 0. TOC 乱码检测
    let toc_garbled = toc_items.is_some_and(toc_items_look_garbled);

    // 1. LLM book-type 校验：Rust 端 LLM 客户端尚未接入主入口（FNM_PHASE12_AUDIT G5）。
    //    config.skip_llm_verify=false 时显式 bail 防误用（AGENTS.md §9）。
    if !config.skip_llm_verify {
        anyhow::bail!(
            "Phase1Config::skip_llm_verify=false 暂不支持——\
             LLM book-type 校验需 vision client 接入主入口（FNM_PHASE12_AUDIT G5）"
        );
    }

    // 2. build_page_partitions —— 把 manual_page_overrides 真传下去（原 None 静默忽略 F7）
    let partitions_result = build_page_partitions(
        pages,
        config.manual_page_overrides.as_ref(),
        config.endnotes_start_page,
    );
    let page_partitions = partitions_result.partitions;

    // 3. 构建 heading candidates（page_rows → collect → normalize）
    let page_rows = crate::chapter_skeleton::heading_candidates::page_rows::legacy_page_rows(
        &page_partitions,
        Some(pages),
    );
    let heading_candidates: Vec<HeadingCandidate> =
        crate::chapter_skeleton::heading_candidates::collect_heading_candidate_rows(
            &page_rows,
            toc_items,
            0,
            &config.pdf_path.clone().unwrap_or_default(),
            None,
            Some(&partitions_result.file_idx_map),
            &config.doc_id.clone().unwrap_or_default(),
        );

    // 4. build_heading_graph（完整版：local_exact + expanded_exact 锚点解析）
    let toc_exportable: Vec<(String, i64, String)> = toc_items
        .map(|items| {
            items
                .iter()
                .filter(|item| item.export_candidate.unwrap_or(true))
                .map(|item| {
                    (
                        item.title.clone(),
                        item.target_pdf_page.unwrap_or(0),
                        item.role_hint.clone(),
                    )
                })
                .collect()
        })
        .unwrap_or_default();
    let page_role_pairs: Vec<(i64, String)> = page_partitions
        .iter()
        .map(|p| (p.page_no, p.page_role.as_str().to_string()))
        .collect();
    let heading_graph = build_heading_graph(&toc_exportable, &heading_candidates, &page_role_pairs);

    // 6. build_chapter_skeleton
    let skeleton = build_chapter_skeleton(
        pages,
        toc_items,
        &page_partitions,
        &heading_graph,
        heading_candidates,
    );

    // 注：`build_book_note_profile` 不在主入口调用——
    // 它推断的 `chapter_modes` 在 phase2 `chapter_split` 内会重新生成
    // （CLAUDE.md §12 分类源头唯一：note_mode 是 phase2 决策权）。
    // phase1 的 book_note_type 模块保留给 LLM book-type verify 用作 prior
    // （FNM_PHASE12_AUDIT G5）。

    // 7. build_section_heads（传入 toc_semantics 输出的 section_heads 作为 fallback sections）
    let fallback_secs: Vec<serde_json::Value> = skeleton
        .toc_semantics_section_heads
        .iter()
        .map(|sh| {
            serde_json::json!({
                "chapter_id": sh.chapter_id,
                "title": sh.title,
                "page_no": sh.page_no,
                "level": sh.level,
                "source": sh.source,
            })
        })
        .collect();
    let (section_heads, _) = build_section_heads(
        &skeleton.chapters,
        &skeleton.heading_candidates,
        &page_partitions,
        Some(&fallback_secs),
    );

    // 8. 构建 overrides_used（←→ Python `build_toc_structure` 行 494-517）
    let mut overrides_used: Vec<OverrideRecord> = Vec::new();
    if let Some(ref overrides) = config.manual_page_overrides {
        for page_str in overrides.keys() {
            overrides_used.push(OverrideRecord {
                kind: "page_override".to_string(),
                description: format!("manual page override for page {}", page_str),
            });
        }
    }

    // 9. 构建 chapter_source_summary 诊断
    let visual_toc_count = skeleton
        .chapters
        .iter()
        .filter(|c| c.source == ChapterSource::VisualToc)
        .count();
    let fallback_count = skeleton
        .chapters
        .iter()
        .filter(|c| c.source == ChapterSource::Fallback)
        .count();
    let chapter_source = if visual_toc_count > 0 {
        "visual_toc"
    } else {
        "fallback"
    };

    // 10. 组装 Phase1Structure（page_partitions 已 owned，零 clone）
    // builder.rs 把 toc_semantics 的完整 meta 写入 diagnostics 顶层，不再嵌套 toc_semantic_meta
    let toc_tree = build_toc_tree(
        toc_items.unwrap_or(&[]),
        &skeleton.chapters,
        &skeleton.diagnostics,
    );

    // D5: TOC tree 过滤/补入（←→ Python toc_structure.py:401-449）
    let mut filtered_chapters = skeleton.chapters.clone();
    if !toc_tree.is_empty() {
        // 构建非章 TOC node 的 title key set
        let toc_non_chapter_keys: std::collections::HashSet<String> = toc_tree
            .iter()
            .filter(|n| !matches!(n.role.as_str(), "chapter" | "container" | "endnotes"))
            .map(|n| fnm_core::title::chapter_title_match_key(&n.title))
            .filter(|k| !k.is_empty())
            .collect();
        // 从章节列表剔除匹配的条目
        if !toc_non_chapter_keys.is_empty() {
            filtered_chapters.retain(|ch| {
                !toc_non_chapter_keys.contains(&fnm_core::title::chapter_title_match_key(&ch.title))
            });
        }
        // 补入 post_body / back_matter 条目
        let existing_keys: std::collections::HashSet<String> = filtered_chapters
            .iter()
            .map(|ch| fnm_core::title::chapter_title_match_key(&ch.title))
            .filter(|k| !k.is_empty())
            .collect();
        let mut extra_nodes: Vec<&fnm_core::records::TocNode> = toc_tree
            .iter()
            .filter(|n| n.role == "post_body" || n.role == "back_matter")
            .collect();
        extra_nodes.sort_by(|a, b| {
            let pa = if a.target_pdf_page > 0 {
                a.target_pdf_page
            } else {
                9999
            };
            let pb = if b.target_pdf_page > 0 {
                b.target_pdf_page
            } else {
                9999
            };
            pa.cmp(&pb).then(a.title.cmp(&b.title))
        });
        let total_pages = page_partitions.len().max(1) as i64;
        let all_starts: Vec<i64> = {
            let mut s: Vec<i64> = filtered_chapters
                .iter()
                .filter_map(|ch| {
                    if ch.start_page > 0 {
                        Some(ch.start_page)
                    } else {
                        None
                    }
                })
                .collect();
            s.sort();
            s
        };
        for node in &extra_nodes {
            let tk = fnm_core::title::chapter_title_match_key(&node.title);
            if tk.is_empty() || existing_keys.contains(&tk) {
                continue;
            }
            let start = if node.target_pdf_page > 0 {
                node.target_pdf_page
            } else {
                0
            };
            let next = all_starts
                .iter()
                .find(|&&s| s > start)
                .copied()
                .unwrap_or(total_pages + 1);
            let end = next - 1;
            let pages: Vec<i64> = if start > 0 && end >= start {
                (start..=end).collect()
            } else {
                vec![]
            };
            filtered_chapters.push(fnm_core::records::ChapterRecord {
                chapter_id: format!(
                    "toc-ch-{:03}-{}",
                    filtered_chapters.len() + 1,
                    &tk[..tk.len().min(20)]
                ),
                title: node.title.clone(),
                start_page: start,
                end_page: end,
                pages,
                source: fnm_core::types::ChapterSource::VisualToc,
                boundary_state: fnm_core::types::BoundaryState::Ready,
            });
        }
    }

    let page_roles = build_page_roles(&page_partitions, &filtered_chapters, 0);

    // 构建 gate_report（在 Phase1Structure 消耗数据前计算）
    let pages_classified = page_partitions.iter().all(|p| {
        let role = p.page_role.as_str();
        !matches!(role, "unknown" | "noise")
    });
    let toc_semantic_contract_ok = skeleton
        .diagnostics
        .get("toc_semantic_contract_ok")
        .and_then(|v| v.as_bool())
        .unwrap_or(true);
    let first_back_matter_page = page_roles
        .iter()
        .filter(|pr| pr.role == "back_matter")
        .map(|pr| pr.page_no)
        .min()
        .unwrap_or(0);
    let no_chapter_after_back_matter = first_back_matter_page <= 0
        || !page_roles
            .iter()
            .any(|pr| pr.role == "chapter" && pr.page_no > first_back_matter_page);
    let role_semantics_valid = toc_semantic_contract_ok && no_chapter_after_back_matter;

    let toc_chapter_order_monotonic = if !toc_tree.is_empty() {
        let chapter_targets: Vec<i64> = toc_tree
            .iter()
            .filter(|n| n.target_pdf_page > 0 && (n.role == "chapter" || n.role == "post_body"))
            .map(|n| n.target_pdf_page)
            .collect();
        if chapter_targets.len() >= 2 {
            chapter_targets.windows(2).all(|w| w[0] <= w[1])
        } else {
            let chs: Vec<i64> = filtered_chapters.iter().map(|c| c.start_page).collect();
            chs.windows(2).all(|w| w[0] <= w[1])
        }
    } else {
        toc_items.map_or(true, |items| {
            let mut prev_page: Option<i64> = None;
            items
                .iter()
                .filter(|t| t.role_hint != "container")
                .all(|t| {
                    let ok =
                        prev_page.map_or(true, |p| t.target_pdf_page.map_or(true, |tp| tp >= p));
                    prev_page = t.target_pdf_page;
                    ok
                })
        })
    };
    let has_exportable_chapters = !filtered_chapters.is_empty();
    let chapter_titles_aligned = skeleton
        .diagnostics
        .get("chapter_title_alignment_ok")
        .and_then(|v| v.as_bool())
        .unwrap_or(true);

    let section_alignment_warn = chapter_titles_aligned;
    let visual_toc_conflict_count = skeleton
        .diagnostics
        .get("visual_toc_conflict_count")
        .and_then(|v| v.as_i64())
        .unwrap_or(0);
    let visual_toc_conflict_warn = visual_toc_conflict_count == 0;

    let gate_report = GateReport {
        hard: HashMap::from([
            ("toc.pages_classified".into(), pages_classified),
            (
                "toc.chapter_order_monotonic".into(),
                toc_chapter_order_monotonic,
            ),
            ("toc.role_semantics_valid".into(), role_semantics_valid),
            (
                "toc.has_exportable_chapters".into(),
                has_exportable_chapters,
            ),
            ("toc.chapter_titles_aligned".into(), chapter_titles_aligned),
        ]),
        soft: HashMap::from([
            ("toc.section_alignment_warn".into(), section_alignment_warn),
            (
                "toc.visual_toc_conflict_warn".into(),
                visual_toc_conflict_warn,
            ),
        ]),
    };

    let structure = Phase1Structure {
        pages: page_partitions,
        heading_candidates: skeleton.heading_candidates,
        chapters: filtered_chapters,
        section_heads,
        endnote_explorer_hints: Default::default(),
        summary: Default::default(),
        toc_tree,
        page_roles,
    };

    Ok(Phase1Output {
        structure,
        diagnostics: serde_json::json!({
            "total_pages": total_pages,
            "toc_garbled": toc_garbled,
            "chapter_source_summary": {
                "source": chapter_source,
                "visual_toc_chapter_count": visual_toc_count,
                "fallback_chapter_count": fallback_count,
            },
        }),
        gate_report,
        overrides_used,
    })
}

/// 把 Phase1Output 持久化到 DB（消耗 ownership，零 clone）。
pub fn persist_phase1(
    repo: &dyn Repository,
    doc_id: &str,
    output: Phase1Output,
) -> anyhow::Result<()> {
    repo.replace_fnm_phase1_products(
        doc_id,
        &fnm_core::db::Phase1Products {
            pages: output.structure.pages,
            chapters: output.structure.chapters,
            heading_candidates: output.structure.heading_candidates,
            section_heads: output.structure.section_heads,
        },
    )?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::input::RawPage;

    #[test]
    fn build_structure_empty_pages() {
        let config = Phase1Config::default();
        let output = build_phase1_structure(&[], None, &config).unwrap();
        assert!(output.structure.pages.is_empty());
        assert!(output.structure.chapters.is_empty());
    }

    #[test]
    fn build_structure_single_page() {
        let pages = vec![RawPage {
            book_page: 1,
            markdown: "# Test\nContent".into(),
            ..Default::default()
        }];
        let config = Phase1Config::default();
        let output = build_phase1_structure(&pages, None, &config).unwrap();
        assert_eq!(output.structure.pages.len(), 1);
        assert_eq!(output.structure.pages[0].page_no, 1);
    }
}
